server/src/test/java/org/elasticsearch/cluster/routing/allocation/SameShardRoutingTests.java

```diff
@@ -137,6 +137,99 @@ public void testSameHost() {
         }
     }
 
+    public void testSameHostNameDifferentHostAddress() {
+        AllocationService strategy = createAllocationService(
+            Settings.builder()
+                .put(SameShardAllocationDecider.CLUSTER_ROUTING_ALLOCATION_SAME_HOST_SETTING.getKey(), true)
+                .put(SameShardAllocationDecider.CLUSTER_ROUTING_ALLOCATION_SAME_HOST_ALWAYS_CHECK_HOSTNAME.getKey(), true)
```

Comment:
Could we also have a test that demonstrates today's behaviour with nodes of different addresses but the same hostname?

Reply:
Thank you again for your comment, and I'm sorry again for the force-push. I added test for scenario that nodes have different addresses but the same hostname with checkHostName off.